File: LINKED LISTS/Insert Operations/InsertAtEnd.java
public class InsertAtEnd {

    // ==============================
    // NODE CLASS
    // ==============================

    static class Node {
        int data;
        Node next;
    }


    // ==============================
    // CREATE A NEW NODE
    // ==============================

    static Node createNode(int data) {
        Node newNode = new Node();
        newNode.data = data;
        newNode.next = null;
        return newNode;
    }


    // ==============================
    // TRAVERSE THE LIST
    // ==============================

    static void traverse(Node head) {
        Node current = head;

        while (current != null) {
            System.out.print(current.data + " ");
            current = current.next;
        }

        System.out.println();
    }


    // ==============================
    // INSERT AT THE END
    // ==============================

    static Node insertAtEnd(Node head, int data) {

        Node newNode = createNode(data);

        /*
            There are two cases.

            ─────────────────────────────────────────
            CASE 1: The list is empty
            ─────────────────────────────────────────

            head
             ↓
            null

            There is no existing node, so the new node
            simply becomes the first (and only) node:

                head = newNode;

            Result:

            head
             ↓
            [newNode] → null


            ─────────────────────────────────────────
            CASE 2: The list is NOT empty
            ─────────────────────────────────────────

            head
             ↓
            [10] → [20] → [30] → null

            We need to find the LAST node — the one
            whose .next is null.

            We use a pointer called "current" and walk
            it forward until current.next == null.

            Why current.next == null and NOT current == null?

            If we stopped at current == null, current
            would have walked PAST the last node.
            We would have no way to attach the new node
            because we lost the reference to [30].

            We stop at current.next == null so that
            current is still pointing AT [30], and then:

                current.next = newNode;

            gives us:

            [10] → [20] → [30] → [newNode] → null
        */

        // CASE 1: empty list
        if (head == null) {
            head = newNode;
            return head;
        }

        // CASE 2: walk to the last node
        Node current = head;

        while (current.next != null) {   // stop BEFORE falling off the end
            current = current.next;
        }

        // current is now the last node — attach the new node
        current.next = newNode;

        return head;
    }


    // ==============================
    // MAIN
    // ==============================

    public static void main(String[] args) {

        /*
            Create the initial list:

            10 → 20 → 30 → null
        */

        Node head = createNode(10);
        Node node2 = createNode(20);
        Node node3 = createNode(30);

        head.next = node2;
        node2.next = node3;

        System.out.print("Original list:       ");
        traverse(head);

        /*
            Insert 40 at the end.

            Before:   10 → 20 → 30
                                  ↓
            After:    10 → 20 → 30 → 40
        */

        head = insertAtEnd(head, 40);

        System.out.print("After inserting 40:  ");
        traverse(head);

        /*
            Insert 50 at the end.

            Before:   10 → 20 → 30 → 40
                                        ↓
            After:    10 → 20 → 30 → 40 → 50
        */

        head = insertAtEnd(head, 50);

        System.out.print("After inserting 50:  ");
        traverse(head);

        /*
            No need to free memory in Java.
            The garbage collector handles it automatically.
        */
    }
}

File: LINKED LISTS/Insert Operations/InsertAtEnd.c
#include <stdio.h>
#include <stdlib.h>


// Defines the structure of a node
struct Node {
    int data;
    struct Node *next;
};


// Creates a new node
struct Node *createNode(int data) {

    struct Node *newNode = malloc(sizeof(struct Node));

    if (newNode == NULL) {
        return NULL;
    }

    newNode->data = data;
    newNode->next = NULL;

    return newNode;
}


// Traverses the list and prints each node's data
void traverse(struct Node *head) {

    struct Node *current = head;

    while (current != NULL) {

        printf("%d ", current->data);

        current = current->next;
    }

    printf("\n");
}


// Inserts a new node at the end of the list
struct Node *insertAtEnd(struct Node *head, int data) {

    // Create the new node
    struct Node *newNode = createNode(data);

    if (newNode == NULL) {

        printf("Memory allocation failed.\n");

        return head;
    }


    /*
        There are two cases.

        CASE 1: The list is empty

        head
         ↓
        NULL

        Since there is no existing node, the new node
        becomes the first node.

            head = newNode;


        CASE 2: The list is not empty

        head
         ↓
        [10] → [20] → [30] → NULL

        We need to find the LAST node.

        Therefore, we stop when:

            current->next == NULL

        Notice that we do NOT use:

            current == NULL

        because that would move current past the last node.

        We need current to remain pointing at [30] so
        that we can change:

            [30].next

        from:

            NULL

        to:

            newNode

        giving:

        [10] → [20] → [30] → [40] → NULL
    */


    // CASE 1: The list is empty
    if (head == NULL) {

        head = newNode;

        return head;
    }


    // CASE 2: Find the last node
    struct Node *current = head;

    while (current->next != NULL) {

        current = current->next;
    }


    // Connect the last node to the new node
    current->next = newNode;


    return head;
}


int main() {

    /*
        Create our initial list:

        10 → 20 → 30 → NULL
    */

    struct Node *head = createNode(10);

    struct Node *node2 = createNode(20);

    struct Node *node3 = createNode(30);


    // Connect the nodes
    head->next = node2;

    node2->next = node3;


    // Display the original list
    printf("Original list: ");

    traverse(head);


    /*
        Insert 40 at the end:

        Before:

        10 → 20 → 30 → NULL

        After:

        10 → 20 → 30 → 40 → NULL
    */

    head = insertAtEnd(head, 40);


    printf("After inserting 40: ");

    traverse(head);


    /*
        Insert another node at the end.
    */

    head = insertAtEnd(head, 50);


    printf("After inserting 50: ");

    traverse(head);


    /*
        Free all dynamically allocated nodes.

        We use a temporary pointer so that we can
        remember where the next node is before
        freeing the current node.
    */

    struct Node *current = head;

    while (current != NULL) {

        struct Node *temp = current;

        current = current->next;

        free(temp);
    }


    return 0;
}